registry/txt_test.go

```diff
@@ -1160,13 +1161,93 @@ func TestExtractRecordType(t *testing.T) {
 	}
 	for _, tc := range tests {
 		t.Run(tc.input, func(t *testing.T) {
-			actualName, actualType := extractRecordType(tc.input)
+			actualName, actualType := extractRecordTypeDefaultPosition(tc.input)
 			assert.Equal(t, tc.expectedName, actualName)
 			assert.Equal(t, tc.expectedType, actualType)
 		})
 	}
 }
 
+func TestToEndpointNameNewTXT(t *testing.T) {
+	type testCase struct {
+		name       string
+		mapper     affixNameMapper
+		domain     string
+		recordType string
+	}
+
+	testCases := []testCase{
+		{
+			name:       "prefix",
+			mapper:     newaffixNameMapper("foo", "", ""),
+			domain:     "example.com",
+			recordType: "A",
+		},
+		{
+			name:       "suffix",
+			mapper:     newaffixNameMapper("", "foo", ""),
+			domain:     "example.com",
+			recordType: "AAAA",
+		},
+		{
+			name:       "prefix with dash",
+			mapper:     newaffixNameMapper("foo-", "", ""),
+			domain:     "example.com",
+			recordType: "A",
+		},
+		{
+			name:       "suffix with dash",
+			mapper:     newaffixNameMapper("", "-foo", ""),
+			domain:     "example.com",
+			recordType: "CNAME",
+		},
+		{
+			name:       "prefix with dot",
+			mapper:     newaffixNameMapper("foo.", "", ""),
+			domain:     "example.com",
+			recordType: "CNAME",
+		},
+		{
+			name:       "suffix with dot",
+			mapper:     newaffixNameMapper("", ".foo", ""),
+			domain:     "example.com",
+			recordType: "CNAME",
+		},
+		{
+			name:       "templated prefix",
+			mapper:     newaffixNameMapper("%{record_type}-foo", "", ""),
+			domain:     "example.com",
+			recordType: "A",
+		},
+		{
+			name:       "templated suffix",
+			mapper:     newaffixNameMapper("", "foo-%{record_type}", ""),
+			domain:     "example.com",
+			recordType: "A",
+		},
+		{
+			name:       "templated prefix with dot",
+			mapper:     newaffixNameMapper("%{record_type}foo.", "", ""),
+			domain:     "example.com",
+			recordType: "CNAME",
+		},
+		{
+			name:       "templated suffix with dot",
+			mapper:     newaffixNameMapper("", ".foo%{record_type}", ""),
+			domain:     "example.com",
+			recordType: "A",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			txtName := tc.mapper.toNewTXTName(tc.domain, tc.recordType)
+			res, _ := tc.mapper.toEndpointName(txtName)
+			assert.Equal(t, tc.domain, res)
```

Comment:
I think this would be a better test if it also asserted the result was a particular expected value, not just that toNewTXTName and toEndpointName are consistent.

Reply:
There's already tests like that, including the one you've referred in here https://github.com/kubernetes-sigs/external-dns/pull/3724#discussion_r1243145276. Risk of these is that they work in isolation but you end up in a situation same as present, where stuff clearly doesn't work.

Reply:
The existing assertions have some cases only covered by `ApplyChanges`; the coverage for `Records` is thin. This will matter once we move to reading but not generating old-format records.

Reply:
We could probably use test cases for prefixes that have two `.` characters.

Reply:
added

Reply:
it now asserts both:
https://github.com/kubernetes-sigs/external-dns/blob/1ffcc5d88cea22a3b9b7acfdaea1c5f66fe5fd9d/registry/txt_test.go#L1285-L1299